File: src/rank_anything/importers.py
"""Turn plain .csv / .tsv / .txt files of numbers into ``samples`` distributions."""

from __future__ import annotations

import csv
import io
import json
import re
from pathlib import Path
from typing import Optional, Union

from .core import DistributionError, parse_number

TABLE_SUFFIXES = (".csv", ".tsv")
TEXT_SUFFIXES = (".txt",)
IMPORT_SUFFIXES = TABLE_SUFFIXES + TEXT_SUFFIXES

# "85,000" or "$1,250,000.50k": commas here are thousands separators, not delimiters.
_THOUSANDS = re.compile(r"^[^\d,]*\d{1,3}(,\d{3})+(\.\d+)?[^\d,]*$")


def _try_number(cell: str) -> Optional[float]:
    try:
        return parse_number(cell)
    except DistributionError:
        return None


def numbers_from_text(text: str) -> list[float]:
    """Numbers from free-form text: one per line, or separated by spaces, tabs,
    semicolons or commas. ``#`` starts a comment. Thousands separators like
    ``85,000`` are understood."""
    out = []
    for lineno, line in enumerate(text.splitlines(), 1):
        line = line.split("#", 1)[0]
        for token in re.split(r"[\s;]+", line):
            parts = [token] if _THOUSANDS.match(token) else token.split(",")
            for part in filter(None, (p.strip() for p in parts)):
                value = _try_number(part)
                if value is None:
                    raise DistributionError(f"line {lineno}: {part!r} is not a number")
                out.append(value)
    return out


def numbers_from_table(text: str, column: Optional[str] = None, delimiter: Optional[str] = None) -> list[float]:
    """Numbers from one column of a CSV/TSV. ``column`` is a header name or a
    1-based index. With no ``column``, the only numeric column is used. A
    non-numeric first row is treated as a header; blank cells are skipped."""
    if delimiter is None:
        try:
            delimiter = csv.Sniffer().sniff(text[:4096], delimiters=",\t;|").delimiter
        except csv.Error:
            delimiter = ","
    rows = [r for r in csv.reader(io.StringIO(text), delimiter=delimiter) if any(c.strip() for c in r)]
    if not rows:
        raise DistributionError("file is empty")

    width = max(len(r) for r in rows)
    first_is_header = any(c.strip() and _try_number(c) is None for c in rows[0])
    header = [c.strip() for c in rows[0]] + [""] * (width - len(rows[0])) if first_is_header else []
    body = rows[1:] if first_is_header else rows
    names = [h or f"column {i + 1}" for i, h in enumerate(header or [""] * width)]

    def cells(i: int) -> list[str]:
        return [r[i].strip() for r in body if i < len(r) and r[i].strip()]

    if column is not None:
        if column in header:
            idx = header.index(column)
        elif column.isdigit() and 1 <= int(column) <= width:
            idx = int(column) - 1
        else:
            raise DistributionError(f"no column {column!r}. Columns: {', '.join(names)}")
    else:
        numeric = [i for i in range(width) if cells(i) and all(_try_number(c) is not None for c in cells(i))]
        if len(numeric) != 1:
            found = ", ".join(names[i] for i in numeric) if numeric else "none"
            raise DistributionError(
                f"can't tell which column to use (numeric columns: {found}). "
                f"Pick one with --column. Columns: {', '.join(names)}"
            )
        idx = numeric[0]

    values, bad = [], []
    for c in cells(idx):
        v = _try_number(c)
        (bad if v is None else values).append(c if v is None else v)
    if bad:
        raise DistributionError(
            f"column {names[idx]!r} has {len(bad)} non-numeric value(s), e.g. {bad[0]!r}"
        )
    return values


def read_numbers(path: Union[str, Path], column: Optional[str] = None) -> list[float]:
    path = Path(path)
    if not path.exists():
        raise DistributionError(f"No such file: {path}")
    text = path.read_text(encoding="utf-8-sig")
    suffix = path.suffix.lower()
    try:
        if suffix in TABLE_SUFFIXES:
            values = numbers_from_table(text, column, "\t" if suffix == ".tsv" else None)
        elif column is not None:
            raise DistributionError("--column only applies to .csv/.tsv files")
        else:
            values = numbers_from_text(text)
    except DistributionError as e:
        raise DistributionError(f"{path}: {e}") from e
    if len(values) < 2:
        raise DistributionError(f"{path}: need at least 2 numbers, found {len(values)}")
    return values


def samples_spec(
    path: Union[str, Path],
    column: Optional[str] = None,
    name: Optional[str] = None,
    title: Optional[str] = None,
    unit: Optional[str] = None,
    description: Optional[str] = None,
    higher_is_better: bool = True,
) -> dict:
    """Build a ``samples`` JSON spec from a file of numbers."""
    path = Path(path)
    values = read_numbers(path, column)
    spec: dict = {"name": name or re.sub(r"[^a-z0-9]+", "-", path.stem.lower()).strip("-") or "custom"}
    if title:
        spec["title"] = title
    spec["description"] = description or f"Imported from {path.name} ({len(values)} values)"
    if unit:
        spec["unit"] = unit
    if not higher_is_better:
        spec["higher_is_better"] = False
    spec["type"] = "samples"
    spec["data"] = [int(v) if v.is_integer() else v for v in values]
    return spec


def dump_spec(spec: dict) -> str:
    """Pretty JSON, but with a ``samples`` data list kept on a single line."""
    if spec.get("type") != "samples" or not isinstance(spec.get("data"), list):
        return json.dumps(spec, indent=2, ensure_ascii=False) + "\n"
    marker = "__RANK_ANYTHING_DATA__"
    text = json.dumps({**spec, "data": marker}, indent=2, ensure_ascii=False)
    return text.replace(json.dumps(marker), json.dumps(spec["data"])) + "\n"

File: src/importers.ts
/** Turn plain .csv / .tsv / .txt text of numbers into `samples` distributions. */

import { DistributionError, parseNumber } from "./distributions.ts";
import { stringifyJson } from "./json.ts";
import { strRepr } from "./text.ts";
import type { DistributionSpec } from "./spec.ts";

export const TABLE_SUFFIXES = [".csv", ".tsv"] as const;
export const TEXT_SUFFIXES = [".txt"] as const;
export const IMPORT_SUFFIXES: readonly string[] = [...TABLE_SUFFIXES, ...TEXT_SUFFIXES];

// "85,000" or "$1,250,000.50k": commas here are thousands separators, not delimiters.
const THOUSANDS = /^[^\d,]*\d{1,3}(,\d{3})+(\.\d+)?[^\d,]*$/;

function tryNumber(cell: string): number | undefined {
  try {
    return parseNumber(cell);
  } catch (e) {
    if (e instanceof DistributionError) return undefined;
    throw e;
  }
}

/** The last path component: "dir/salaries.csv" -> "salaries.csv". */
export function basename(path: string): string {
  const parts = path.split(/[\\/]/).filter(Boolean);
  return parts[parts.length - 1] ?? "";
}

/** "salaries.csv" -> ".csv" (lowercased); "" when there is no extension. */
export function suffix(path: string): string {
  const name = basename(path);
  const i = name.lastIndexOf(".");
  return i > 0 && i < name.length - 1 ? name.slice(i).toLowerCase() : "";
}

/** "dir/salaries.csv" -> "salaries". */
export function stem(path: string): string {
  const name = basename(path);
  const i = name.lastIndexOf(".");
  return i > 0 && i < name.length - 1 ? name.slice(0, i) : name;
}

const LINE_BREAK = /\r\n|[\n\r\v\f\x1c-\x1e\x85\u2028\u2029]/;

/**
 * Numbers from free-form text: one per line, or separated by spaces, tabs,
 * semicolons or commas. `#` starts a comment. Thousands separators like
 * `85,000` are understood.
 */
export function numbersFromText(text: string): number[] {
  const out: number[] = [];
  const lines = text.split(LINE_BREAK);
  lines.forEach((raw, i) => {
    const line = raw.split("#")[0];
    for (const token of line.split(/[\s;]+/)) {
      const parts = THOUSANDS.test(token) ? [token] : token.split(",");
      for (const part of parts.map((p) => p.trim()).filter(Boolean)) {
        const value = tryNumber(part);
        if (value === undefined) throw new DistributionError(`line ${i + 1}: ${strRepr(part)} is not a number`);
        out.push(value);
      }
    }
  });
  return out;
}

// ------------------------------------------------------------------ CSV

/** Parse CSV text (quoted fields, doubled quotes, any line ending) with the given delimiter. */
export function parseCsv(text: string, delimiter = ","): string[][] {
  const rows: string[][] = [];
  let row: string[] = [];
  let field = "";
  let inQuotes = false;
  let quotedField = false;
  let rowStarted = false;
  for (let i = 0; i < text.length; i++) {
    const c = text[i];
    if (inQuotes) {
      if (c === '"') {
        if (text[i + 1] === '"') {
          field += '"';
          i++;
        } else inQuotes = false;
      } else field += c;
      continue;
    }
    if (c === '"' && field === "" && !quotedField) {
      inQuotes = quotedField = rowStarted = true;
    } else if (c === delimiter) {
      row.push(field);
      field = "";
      quotedField = false;
      rowStarted = true;
    } else if (c === "\n" || c === "\r") {
      if (c === "\r" && text[i + 1] === "\n") i++;
      if (rowStarted || field) row.push(field);
      rows.push(row);
      row = [];
      field = "";
      quotedField = rowStarted = false;
    } else {
      field += c;
      rowStarted = true;
    }
  }
  if (rowStarted || field) {
    row.push(field);
    rows.push(row);
  }
  return rows;
}

const PREFERRED = [",", "\t", ";", " ", ":"];

/** Delimiter guess from the characters next to quoted fields. */
function guessQuoteAndDelimiter(data: string, delimiters: string): string {
  const d = String.raw`[^\p{L}\p{N}_\n"']`;
  const patterns = [
    String.raw`(?<delim>${d})(?<space> ?)(?<quote>["']).*?\k<quote>\k<delim>`,
    String.raw`(?:^|\n)(?<quote>["']).*?\k<quote>(?<delim>${d})(?<space> ?)`,
    String.raw`(?<delim>${d})(?<space> ?)(?<quote>["']).*?\k<quote>(?:$|\n)`,
    String.raw`(?:^|\n)(?<quote>["']).*?\k<quote>(?:$|\n)`,
  ];
  for (const p of patterns) {
    const matches = [...data.matchAll(new RegExp(p, "gmsu"))];
    if (!matches.length) continue;
    const delims = new Map<string, number>();
    for (const m of matches) {
      const key = m.groups?.delim;
      if (key && delimiters.includes(key)) delims.set(key, (delims.get(key) ?? 0) + 1);
    }
    let best = "";
    let count = 0;
    for (const [k, n] of delims) if (n > count) [best, count] = [k, n];
    return best === "\n" ? "" : best;
  }
  return "";
}

/** Delimiter guess: the character that appears equally often on (almost) every line. */
function guessDelimiter(sample: string, delimiters: string): string {
  const data = sample.split("\n").filter(Boolean);
  const chunkLength = Math.min(10, data.length);
  const charFrequency = new Map<string, Map<number, number>>();
  const delims = new Map<string, [number, number]>();
  let iteration = 0;
  for (let start = 0, end = chunkLength; start < data.length; start = end, end += chunkLength) {
    iteration++;
    for (const line of data.slice(start, end)) {
      for (const ch of delimiters) {
        const meta = charFrequency.get(ch) ?? new Map<number, number>();
        const freq = line.split(ch).length - 1;
        meta.set(freq, (meta.get(freq) ?? 0) + 1);
        charFrequency.set(ch, meta);
      }
    }
    const modes = new Map<string, [number, number]>();
    for (const [ch, meta] of charFrequency) {
      const items = [...meta.entries()];
      if (items.length === 1 && items[0][0] === 0) continue;
      let mode = items[0];
      for (const it of items) if (it[1] > mode[1]) mode = it;
      const rest = items.filter((it) => it !== mode).reduce((s, it) => s + it[1], 0);
      modes.set(ch, [mode[0], mode[1] - rest]);
    }
    const total = Math.min(chunkLength * iteration, data.length);
    for (let consistency = 1.0; delims.size === 0 && consistency >= 0.9; consistency -= 0.01) {
      for (const [k, v] of modes) {
        if (v[0] > 0 && v[1] > 0 && v[1] / total >= consistency) delims.set(k, v);
      }
    }
    if (delims.size === 1) return [...delims.keys()][0];
  }
  if (!delims.size) return "";
  for (const d of PREFERRED) if (delims.has(d)) return d;
  const items = [...delims.entries()].sort(
    ([ka, a], [kb, b]) => a[0] - b[0] || a[1] - b[1] || (ka < kb ? -1 : ka > kb ? 1 : 0),
  );
  return items[items.length - 1][0];
}

/** Guess a CSV delimiter among `delimiters`; "," if unsure. */
export function sniffDelimiter(sample: string, delimiters = ",\t;|"): string {
  return guessQuoteAndDelimiter(sample, delimiters) || guessDelimiter(sample, delimiters) || ",";
}

/**
 * Numbers from one column of a CSV/TSV. `column` is a header name or a
 * 1-based index. With no `column`, the only numeric column is used. A
 * non-numeric first row is treated as a header; blank cells are skipped.
 */
export function numbersFromTable(text: string, column?: string | null, delimiter?: string | null): number[] {
  const delim = delimiter ?? sniffDelimiter(text.slice(0, 4096));
  const rows = parseCsv(text, delim).filter((r) => r.some((c) => c.trim()));
  if (!rows.length) throw new DistributionError("file is empty");

  const width = Math.max(...rows.map((r) => r.length));
  const firstIsHeader = rows[0].some((c) => c.trim() && tryNumber(c) === undefined);
  const header = firstIsHeader
    ? [...rows[0].map((c) => c.trim()), ...Array<string>(width - rows[0].length).fill("")]
    : [];
  const body = firstIsHeader ? rows.slice(1) : rows;
  const names = (header.length ? header : Array<string>(width).fill("")).map((h, i) => h || `column ${i + 1}`);
  const cells = (i: number) => body.filter((r) => i < r.length && r[i].trim()).map((r) => r[i].trim());

  let idx: number;
  if (column != null) {
    if (header.includes(column)) idx = header.indexOf(column);
    else if (/^\d+$/.test(column) && Number(column) >= 1 && Number(column) <= width) idx = Number(column) - 1;
    else throw new DistributionError(`no column ${strRepr(column)}. Columns: ${names.join(", ")}`);
  } else {
    const numeric = [...Array(width).keys()].filter((i) => {
      const cs = cells(i);
      return cs.length > 0 && cs.every((c) => tryNumber(c) !== undefined);
    });
    if (numeric.length !== 1) {
      const found = numeric.length ? numeric.map((i) => names[i]).join(", ") : "none";
      throw new DistributionError(
        `can't tell which column to use (numeric columns: ${found}). Pick one with --column. Columns: ${names.join(", ")}`,
      );
    }
    idx = numeric[0];
  }

  const values: number[] = [];
  const bad: string[] = [];
  for (const c of cells(idx)) {
    const v = tryNumber(c);
    if (v === undefined) bad.push(c);
    else values.push(v);
  }
  if (bad.length) {
    throw new DistributionError(`column ${strRepr(names[idx])} has ${bad.length} non-numeric value(s), e.g. ${strRepr(bad[0])}`);
  }
  return values;
}

/**
 * Numbers from the contents of a .csv/.tsv/.txt file. `filename` picks the
 * format (by extension) and prefixes error messages.
 */
export function numbersFromFile(text: string, filename: string, column?: string | null): number[] {
  if (text.startsWith("﻿")) text = text.slice(1);
  const ext = suffix(filename);
  let values: number[];
  try {
    if ((TABLE_SUFFIXES as readonly string[]).includes(ext)) {
      values = numbersFromTable(text, column, ext === ".tsv" ? "\t" : null);
    } else if (column != null) {
      throw new DistributionError("--column only applies to .csv/.tsv files");
    } else {
      values = numbersFromText(text);
    }
  } catch (e) {
    if (e instanceof DistributionError) throw new DistributionError(`${filename}: ${e.message}`);
    throw e;
  }
  if (values.length < 2) throw new DistributionError(`${filename}: need at least 2 numbers, found ${values.length}`);
  return values;
}

export interface SamplesSpecOptions {
  /** CSV column: header name or 1-based index. Default: the only numeric column. */
  column?: string | null;
  name?: string | null;
  title?: string | null;
  unit?: string | null;
  description?: string | null;
  higherIsBetter?: boolean;
}

/** Build a `samples` JSON spec from the contents of a .csv/.tsv/.txt file of numbers. */
export function samplesSpecFromText(text: string, filename: string, options: SamplesSpecOptions = {}): DistributionSpec {
  const values = numbersFromFile(text, filename, options.column);
  const name = options.name || stem(filename).toLowerCase().replace(/[^a-z0-9]+/g, "-").replace(/^-+|-+$/g, "") || "custom";
  const spec: Record<string, unknown> = { name };
  if (options.title) spec.title = options.title;
  spec.description = options.description || `Imported from ${basename(filename)} (${values.length} values)`;
  if (options.unit) spec.unit = options.unit;
  if (options.higherIsBetter === false) spec.higher_is_better = false;
  spec.type = "samples";
  spec.data = values;
  return spec as unknown as DistributionSpec;
}

/** Pretty JSON, but with a `samples` data list kept on a single line. */
export function dumpSpec(spec: object): string {
  const s = spec as { type?: unknown; data?: unknown };
  const inline = s.type === "samples" && Array.isArray(s.data) ? ["data"] : [];
  return stringifyJson(spec, 2, inline) + "\n";
}
